C++ ClangCL on Windows - `/wd4668` flag results in error

I'm compiling a C++ project on Windows using ClangCL from Scoop's `versions/mingw-winlibs-llvm` package. In my CMakeLists.txt file I call `set(CMAKE_CXX_COMPILER_LAUNCHER "${SCCACHE_PROGRAM}")`, and I can compile my project with MSVC, Clang, GCC and ClangCL with Visual Studio generator without having to install the Clang toolset from VS Installer (after setting the `llvm-mt` executable manually before a `project()` call), but trying to compile using ClangCL with the MinGW Makefiles generator option in CLion results in:
```
clang-cl: error: no such file or directory: '/wd4668'
mingw32-make[3]: *** [_deps\catch2-build\src\CMakeFiles\Catch2.dir\build.make:74: _deps/catch2-build/src/CMakeFiles/Catch2.dir/catch2/benchmark/catch_chronometer.cpp.obj] Error 1
mingw32-make[3]: *** Waiting for unfinished jobs....
```
This does not happen when using `ccache` instead of `sccache`.
I've installed `ccache` and `sccache` using Scoop as well.

Correct me if any of the mentioned steps are incorrect or I'm doing something wrong 💯

I think the error is unrelated, but the fix is what's causing the issue (correct me if I'm wrong); I had no issues with the warning that the original PR references (`error C4668: '__midl' is not defined as a preprocessor macro, replacing with '0' for '#if/#elif'`). I'm running version `0.5.4` right now which has that fix included.

I am new to rust. In the project, it is said that the Windows SDK preprocessing stage will generate C4668. After testing, the clang-cl preprocessing stage does not seem to generate C4668, so your error will appear. The solution is to recompile or add codes such as macro definitions to control the addition of wd4668 errors during the use of clang-cl.

> 我正在使用 Scoop 包中的 ClangCL 在 Windows 上编译 C++ 项目`versions/mingw-winlibs-llvm`。在我的 CMakeLists.txt 文件中，我调用`set(CMAKE_CXX_COMPILER_LAUNCHER "${SCCACHE_PROGRAM}")`，我可以使用 Visual Studio 生成器使用 MSVC、Clang、GCC 和 ClangCL 编译我的项目，而无需从 VS Installer 安装 Clang 工具集（在调用之前手动设置可执行文件之后），但尝试使用 ClangCL 和 CLion 中的 MinGW Makefiles 生成器选项进行编译会`llvm-mt`导致`project()`：
> 
> ```
> clang-cl: error: no such file or directory: '/wd4668'
> mingw32-make[3]: *** [_deps\catch2-build\src\CMakeFiles\Catch2.dir\build.make:74: _deps/catch2-build/src/CMakeFiles/Catch2.dir/catch2/benchmark/catch_chronometer.cpp.obj] Error 1
> mingw32-make[3]: *** Waiting for unfinished jobs....
> ```
> 
> `ccache`使用代替 时不会发生这种情况`sccache`。 我也安装`ccache`并`sccache`使用了 Scoop。
> 
> 如果上述任何步骤不正确或我做错了什么，请纠正我💯

I am new to rust. In the project, it is said that the Windows SDK preprocessing stage will generate C4668. After testing, the clang-cl preprocessing stage does not seem to generate C4668, so your error will appear. The solution is to recompile or add codes such as macro definitions to control the addition of wd4668 errors during the use of clang-cl.

> I'm compiling a C++ project on Windows using ClangCL from Scoop's `versions/mingw-winlibs-llvm` package. In my CMakeLists.txt file I call `set(CMAKE_CXX_COMPILER_LAUNCHER "${SCCACHE_PROGRAM}")`, and I can compile my project with MSVC, Clang, GCC and ClangCL with Visual Studio generator without having to install the Clang toolset from VS Installer (after setting the `llvm-mt` executable manually before a `project()` call), but trying to compile using ClangCL with the MinGW Makefiles generator option in CLion results in:
> 
> ```
> clang-cl: error: no such file or directory: '/wd4668'
> mingw32-make[3]: *** [_deps\catch2-build\src\CMakeFiles\Catch2.dir\build.make:74: _deps/catch2-build/src/CMakeFiles/Catch2.dir/catch2/benchmark/catch_chronometer.cpp.obj] Error 1
> mingw32-make[3]: *** Waiting for unfinished jobs....
> ```
> 
> This does not happen when using `ccache` instead of `sccache`. I've installed `ccache` and `sccache` using Scoop as well.
> 
> Correct me if any of the mentioned steps are incorrect or I'm doing something wrong 💯

I am new to rust. In the project, it is said that the Windows SDK preprocessing stage will generate C4668. After testing, the clang-cl preprocessing stage does not seem to generate C4668, so your error will appear. The solution is to recompile or add codes such as macro definitions to control the addition of wd4668 errors during the use of clang-cl.

--- a/src/compiler/msvc.rs
+++ b/src/compiler/msvc.rs
@@ -885,9 +885,6 @@ where
         .args(&parsed_args.preprocessor_args)
         .args(&parsed_args.dependency_args)
         .args(&parsed_args.common_args)
-        // Windows SDK generates C4668 during preprocessing, but compiles fine.
-        // Read for more info: https://github.com/mozilla/sccache/issues/1725
-        .arg("/wd4668")
         .env_clear()
         .envs(env_vars.iter().map(|(k, v)| (k, v)))
         .current_dir(cwd);
@@ -902,6 +899,9 @@ where
             cmd.arg("/sourceDependencies");
             cmd.arg(depfile);
         }
+        // Windows SDK generates C4668 during preprocessing, but compiles fine.
+        // Read for more info: https://github.com/mozilla/sccache/issues/1725
+        cmd.arg("/wd4668");
     }
 
     if rewrite_includes_only && is_clang {

